### PR TITLE
Inference fix

### PR DESCRIPTION
**Pull request by YingboMa**

### CODE REVIEW

File: src/ensemble/basic_ensemble_solve.jl [hunk 1]
```diff
@@ -75,36 +75,56 @@ function __solve(prob::AbstractEnsembleProblem,
...
-    _u = map(i->u[i],1:length(u))
-  else
-    _u = u
-  end
-  return EnsembleSolution(_u,elapsed_time,converged)
+
+  u = reduce(vcat, u)
+  _u = tighten_container_eltype(u)
+
+  return EnsembleSolution(u,elapsed_time,converged)
```
Comment (devmotion) [inline src/ensemble/basic_ensemble_solve.jl:124]:
```suggestion
  return EnsembleSolution(_u,elapsed_time,converged)

```

File: test/downstream/inference.jl [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+function lorenz(du,u,p,t)
+ du[1] = 10.0(u[2]-u[1])
+ du[2] = u[1]*(28.0-u[3]) - u[2]
+ du[3] = u[1]*u[2] - (8/3)*u[3]
+end
+u0 = [1.0;0.0;0.0]
+tspan = (0.0,1.0)
+prob = ODEProblem(lorenz,u0,tspan)
+sol = solve(prob,Tsit5())
+@inferred solve(prob,Tsit5())
```
Comment (devmotion) [inline test/downstream/inference.jl:11]: I guess it would make sense to add some ensemble inference tests as well?

File: test/runtests.jl [hunk 3]
```diff
@@ -38,6 +38,7 @@ if !is_APPVEYOR && GROUP == "Downstream"
     @time @safetestset "Null Parameters" begin include("downstream/null_params_test.jl") end
     @time @safetestset "Ensemble Simulations" begin include("downstream/ensemble.jl") end
     @time @safetestset "Ensemble Analysis" begin include("downstream/ensemble_analysis.jl") end
+    @time @safetestset "Inferece Tests" begin include("downstream/inference.jl") end
```
Comment (devmotion) [inline test/runtests.jl:41]:
```suggestion
    @time @safetestset "Inference Tests" begin include("downstream/inference.jl") end
```

File: test/downstream/inference.jl (hunk 2 above)
Comment (ChrisRackauckas) [reply test/downstream/inference.jl:11]: yup was in the middle of pushing 👍

File: src/ensemble/basic_ensemble_solve.jl [hunk 4]
```diff
@@ -75,36 +75,53 @@ function __solve(prob::AbstractEnsembleProblem,
...
+  if num_batches == 1 && prob.reduction === DEFAULT_REDUCTION
+    elapsed_time = @elapsed u = batch_function(1:trajectories)
+    _u = tighten_container_eltype(u)
+    return EnsembleSolution(_u,elapsed_time,true)
+  end
+
+  converged::Bool = false
+
+  batch_data = batch_function(I)
+  u = similar(batch_data, 0)
```
Comment (devmotion) [inline src/ensemble/basic_ensemble_solve.jl:104]: This seems a bit weird, I don't really understand why you need that? Isn't assigning `u` to the initial reduction in the next line sufficient?
```suggestion
```

Comment (YingboMa) [reply src/ensemble/basic_ensemble_solve.jl:104]: The next line uses `u` as well.
`prob.reduction(u,batch_data,I)`

Comment (YingboMa) [inline src/ensemble/basic_ensemble_solve.jl:104]: Ah, it should be
```suggestion
  u = prob.u_init === nothing ? similar(batch_data, 0) : prob.u_init
```

Comment (devmotion) [reply src/ensemble/basic_ensemble_solve.jl]: Yes, I was mainly wondering what the motivation is for not using `u = deepcopy(prob.u_init)` (or something similar) as in the old code but `u = similar(batch_data, 0)` in the initial reduction step.

Comment (YingboMa) [reply src/ensemble/basic_ensemble_solve.jl]: See https://github.com/SciML/DiffEqBase.jl/pull/511#discussion_r426237782

**Comment by ChrisRackauckas** [comment]
https://travis-ci.org/github/SciML/DiffEqBase.jl/jobs/688001753#L1013

ugh, the ensemble tests pass again, but that broke the inference again. I need to catch a few zzzzz's, and can probably return to this in a bit.

**Comment by ChrisRackauckas** [comment]
@YingboMa let's continue this in another PR.

**Pull request merged**